Creation of command "list"

youngmoney opened the pull request:
This work in progress pull request is to illustrate where I am with the list command, currently it supports lots of formatting and grabs the correct events, however there are date/datetime problems:

1.  It cannot accept times (this is crucial for functionality that myself and others want)

2. events need to accept a range of times to fully understand how to format themselves

3. I would like to factor out a lot of date range/datetime if else statements into their own object


This request also contains a change to the event less than function allowing it to work across days

geier commented:
Good to see that you are making progress. Do you need any help?

Perhaps you could split what you have into several patches/PRs which we could integrate faster (and which are also easier to review). `format` would probably be a good candidate for that. Also, please write some (simple) tests for new/changed features and run flake8 over your code.

Youngmoney commented:
I don't think I need any help, I am just busy the next week or so. Hopefully after that I can finish this off. I would like to break it into two different PRs, one for the updates to the even object, which is `format` and `__lt__` and another that would follow that sets up the list command. I noticed that github is saying that I am failing the travis builds, is that something I should look into? nothing I run locally is having any problems I'm aware of

geier commented:
Yes, some tests are failing, which might be expected in an early stage and you probably should run `flake8` over the code. Please have a look at the [documentation](https://khal.readthedocs.io/en/latest/contributing.html#hacking), which is very much work in progress. If something isn't clear or could be more helpful please point it out (or submit a PR).

Youngmoney commented:
so, I have finished all the fixes I wanted to make, now I am trying to prepare this for a pull request. Because my second round of fixes is small, I'm hoping one pull request will the list and format together is fine. I have some question about testing...well more I'm just confused. I don't do any real python work (with tests and all) so I have no idea how this testing suite works. I tried installing tox and running it but that errored out without running any tests.

geier commented:
Sadly, tox doesn't work very reliable for me either. You can install
pytest instead and than run `pytest tests` (if you install from pyPI
it's called `py.test`.

Additionally, install `flake8 --max-line-length=100` and fix any errors.

Quoting Youngmoney (2016-05-07 00:24:29)
> so, I have finished all the fixes I wanted to make, now I am trying to prepare this for a pull request. Because my second round of fixes is small, I'm hoping one pull request will the list and format together is fine. I have some question about testing...well more I'm just confused. I don't do any real python work (with tests and all) so I have no idea how this testing suite works. I tried installing tox and running it but that errored out without running any tests.
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/pimutils/khal/pull/410#issuecomment-217578292

Youngmoney commented:
I'm a bit lost as to why I have triggered so many failed tests, from what I can see (through my diff with master) is that I have only added new self contained functions with the exception of the change of `__lt__` in the event class, and when I change that back I still fail the same tests

geier commented:
I'll have a look later today. 

Am 7. Mai 2016 02:05:34 MESZ, schrieb Youngmoney <notifications@github.com>:
>I'm a bit lost as to why I have triggered so many failed tests, from
>what I can see (through my diff with master) is that I have only added
>new self contained functions with the exception of the change of
>`__lt__` in the event class, and when I change that back I still fail
>the same tests
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/khal/pull/410#issuecomment-217590665

-- 
Sent from my phone. Please excuse my brevity.

Youngmoney commented:
found the problem. I named the command list, `def list()` was causing problems so renamed it. I am now only failing style. Will try to update style and write changelog/authors.txt soon

Youngmoney commented:
I have now updated authors, the changelog, fixed all style, and added a test. Let me know if there is anything else I need to do. I would love to see this in master soon!

geier commented on doc/source/usage.rst:
```diff
@@ -89,6 +89,95 @@ string *now*. ``at`` defaults to *now*.
+
+::
+
+        khal list --format "{title} {description}"
```
This example doesn't work for me, I get:

```
khal list --format "{title} {description}" today                                               ~/workspace/khal
Traceback (most recent call last):
  File "/home/cg/.virtualenvs/khal/bin/khal", line 9, in <module>
    load_entry_point('khal==0.8.2.dev20+ngd64a1e3.d20160425', 'console_scripts', 'khal')()
  File "/usr/home/cg/.virtualenvs/khal/lib/python3.4/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/home/cg/.virtualenvs/khal/lib/python3.4/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/home/cg/.virtualenvs/khal/lib/python3.4/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/home/cg/.virtualenvs/khal/lib/python3.4/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/home/cg/.virtualenvs/khal/lib/python3.4/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/home/cg/.virtualenvs/khal/lib/python3.4/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/home/cg/workspace/khal/khal/cli.py", line 327, in klist
    locale=ctx.obj['conf']['locale'],
  File "/usr/home/cg/workspace/khal/khal/controllers.py", line 263, in khal_list
    end=end, once=once, nostarted=nostarted, **kwargs)
  File "/usr/home/cg/workspace/khal/khal/controllers.py", line 187, in get_list_from_str
    start = aux.guessdatetimefstr(start.split(" "), locale)[0]
  File "/usr/home/cg/workspace/khal/khal/aux.py", line 211, in guessdatetimefstr
    raise ValueError()
ValueError
```

geier commented:
Nice work so far!

Before having a look at the implementation, I do have one conceptual question: what differentiates the new ``list`` command from ``at`` and ``agenda``? I would like to remove (or at least deprecate) at least one of them with the new ``list``.

Also, could you squash this into two or three topical commits, e.g. "\_lt\_ fix", "--format" and "list"?

Youngmoney commented:
the fact that list is failing there is weird, I have it working on my end with a copy-paste of what you put in. does it work if you leave it blank (no "today")?

I believe this can replace both at and list. I just added an option to make that more true: the end argument can now be one of +INT{d,m,h} where +5m means start + 5 minutes, and the same for days or hours. This will allow for at and agenda to be almost the same (from what I understand/use them for) (i pushed this change).

I will rebase later today, I have a weird commit history will pulls in there and some mixed commits

Youngmoney commented:
I'm really very confused as to why its not working for you, I would say try to see why the calls to `aux.guessdatetimefstr` are not accepting today, as it should be the final check in the loop there.

Youngmoney commented:
i have squashed, let me know if you bug is still occurring I can try to look into it on my end, but like I said its working for me so Im a bit confused

geier commented:
Works now, thanks.

Would you say this ``khal list 10.05.2016 12:30 +2d`` is supposed to work?

Youngmoney commented:
this `khal list "10.05.2016 12:30" +2d` should work, I have each date as a separate argument instead of trying to parse it into its parts like agenda does. Im more for this option because it makes more sense in a unix way, i could also see us having a distinct separator such as `..` (which gcalcli uses) to separate start and end, but I think that is a different discussion.

Youngmoney commented:
before you look too much into he implementation I wanted to say that now that I have more time I am going to go through and try to clean it up a bit and also add some format options for calendar and calendar color if possible, as well as a config option for default formatting

Youngmoney commented:
I have done the refactoring I want to do, I looked into the calendar but the names I can get from "event" is the wrong name (not the display name), so I have no added the calendar name.

geier commented:
Here are some notes (in no particular order) on your PR (I cannot give them
inline, as I'm on a train and internet connection is notoriously bad on German
trains). Please note, that while I have a lot of comments and improvement
suggestions, it's because you are touching two very core components of khal,
both of which I would like to become the main ways of interacting with khal.


Conceptual remarks on `list`
============================
  
* with the current implementation we cannot replace list AFAIK, because `khal
  list DTSTART` will always print all events on the day given by DTSTART. Yet
  making `list` print only events which are scheduled for DTSTART unless an
  explicit end datetime is given is ugly from some (my) user scenario
  (but perhaps we should just keep `at` and be a bit less ambitious about
  replacing both `agenda` and `at` with `list`)

* I really would like to see a default timedelta (configurable in the
  configuration file), because a) because my general usage is looking for events
  scheduled for today and tomorrow and I don't wont to type `khal list today
  +1d` every time I could just type `khal list`, and it's what
  calendar(1) does (shows events for today and tomorrow by default)

* `khal list Tomorrow 10:30` should work without quotation marks,
  (and be interpreted as start='tomorrow 10:30') one might want to argue
  that only having optional arguments at the end of an argument list is
  normal UNIX behaviour, but a) I think it's still pretty clear what
  `khal list Tomorrow 10:30` is supposed to mean, b) it's better to be
  consistent than to be right, and c) IMO `khal list 'Today 12:30'
  'Tomorrow 16:30'` is just plain ugly

  In case you, or other people strongly disagree here, we can a) make
  both approaches work or b) introduce (optional) --start DATETIME and
  --end DATETIME options.

* I like the current `agenda` output, where it display's a line for every new
  day (with its date in it), this behaviour could probably be introduced with an
  option, but this is not priority and I'll accept the PR without it

Other remarks
=============

* it's currently too easy to see a stacktrace as a user, please wrap those place
  with a try-except-log-exit, some examples:
  * khal list todayx
  * khal list today 1days
  * khal list today --format '{titlex}'


* regarding: `--nostarted`, shouldn't it be `--notstarted`? And I believe the
  help would be better with an added `only`: "Print only events which have not
  started", btw. I really like that option, I see myself aliasing something like
  `khal list now +1d --notstarted` to `khal next`. Perhaps we can support
  aliases like git does, @untitaker is this something click is currently capable
  of?

* could you use guesstimedeltafstr for timedelta parsing? As both versions are
  probably not optimal yet, I would be a waste to keep improving both

* some properties `.format()` probably should support (but we can always add
  these later):
  * start-long, end-long (not sure about the names)
  * color (the color of the calendar this event belongs to)
  * calendar (even if it's not the display-name yet)

* the 'XXX generate as unicode in the first place' comment can be removed, this
  is unicode in all cases, at the very least since we are python 3 only (I know
  there is another comment like it, but it makes no sense either)

* please also run `pep8 --ignore=E501` on your PR (there are some operators with
  missing whitespace), I'm not sure why flake8 doesn't catch that

* no single line `ifs` please, e.g. if not "seen-days" in env: env["seen-days"] = set()

If you make this two PR (list and format) I believe we can merge format a lot
earlier and can remove `event_description`.

That's it for now...

On Mon, May 16, 2016 at 07:47:59PM -0700, Youngmoney wrote:
> I have done the refactoring I want to do, I looked into the calendar but the names I can get from "event" is the wrong name (not the display name), so I have no added the calendar name.
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/pimutils/khal/pull/410#issuecomment-219606051

Youngmoney commented:
So I think I've fixed these all:

On Tue, May 17, 2016 at 02:38:42AM -0700, Christian Geier wrote:
>Here are some notes (in no particular order) on your PR (I cannot give them
>inline, as I'm on a train and internet connection is notoriously bad on German
>trains). Please note, that while I have a lot of comments and improvement
>suggestions, it's because you are touching two very core components of khal,
>both of which I would like to become the main ways of interacting with khal.
>
>
>Conceptual remarks on `list`
>============================
>
>* with the current implementation we cannot replace list AFAIK, because `khal
>  list DTSTART` will always print all events on the day given by DTSTART. Yet
>  making `list` print only events which are scheduled for DTSTART unless an
>  explicit end datetime is given is ugly from some (my) user scenario
>  (but perhaps we should just keep `at` and be a bit less ambitious about
>  replacing both `agenda` and `at` with `list`)

* In regards to replacing both at and agenda, I think at is a special
  command but I think we could just make it a wrapper for list very
  easily, so at least supporting it would be simple. In addition it
  would be great to see all commands use the list functions.

>
>* I really would like to see a default timedelta (configurable in the
>  configuration file), because a) because my general usage is looking for events
>  scheduled for today and tomorrow and I don't wont to type `khal list today
>  +1d` every time I could just type `khal list`, and it's what
>  calendar(1) does (shows events for today and tomorrow by default)

* timedelta now exists in config under default.timedelta
>
>* `khal list Tomorrow 10:30` should work without quotation marks,
>  (and be interpreted as start='tomorrow 10:30') one might want to argue
>  that only having optional arguments at the end of an argument list is
>  normal UNIX behaviour, but a) I think it's still pretty clear what
>  `khal list Tomorrow 10:30` is supposed to mean, b) it's better to be
>  consistent than to be right, and c) IMO `khal list 'Today 12:30'
>  'Tomorrow 16:30'` is just plain ugly

* daterange is now accepted as a serious of words and they are parsed,
  along with attempting to use the timedelta, in addition the option for
  the end to just be `eod` is available (this is to override any
  timedelta defaults and see normal day behaviour)
>
>  In case you, or other people strongly disagree here, we can a) make
>  both approaches work or b) introduce (optional) --start DATETIME and
>  --end DATETIME options.
>
>* I like the current `agenda` output, where it display's a line for every new
>  day (with its date in it), this behaviour could probably be introduced with an
>  option, but this is not priority and I'll accept the PR without it
>

* newline/header is possible: `--format
  "{start-date-once-newline}{start-date-once}{start-date-once-newline}{title}
  ..."`, however it might be nice in the future to support some sort of
  "pretty date" for this, but that should be in the config as a
  datestring and such and I did not want to complicate the number of
  date options.

>Other remarks
>=============
>
>* it's currently too easy to see a stacktrace as a user, please wrap those place
>  with a try-except-log-exit, some examples:
>  * khal list todayx
>  * khal list today 1days
>  * khal list today --format '{titlex}'
>

* stacktraces are pretty

>
>* regarding: `--nostarted`, shouldn't it be `--notstarted`? And I believe the
>  help would be better with an added `only`: "Print only events which have not
>  started", btw. I really like that option, I see myself aliasing something like
>  `khal list now +1d --notstarted` to `khal next`. Perhaps we can support
>  aliases like git does, @untitaker is this something click is currently capable
>  of?

* nostarted is notstarted
>
>* could you use guesstimedeltafstr for timedelta parsing? As both versions are
>  probably not optimal yet, I would be a waste to keep improving both

* guesstimedeltafstr is used
>
>* some properties `.format()` probably should support (but we can always add
>  these later):
>  * start-long, end-long (not sure about the names)
>  * color (the color of the calendar this event belongs to)
>  * calendar (even if it's not the display-name yet)

* `.format()` supports: calendar, calendar-color, start-long, end-long,
  start-date-long, end-date-long.
>
>* the 'XXX generate as unicode in the first place' comment can be removed, this
>  is unicode in all cases, at the very least since we are python 3 only (I know
>  there is another comment like it, but it makes no sense either)
>
>* please also run `pep8 --ignore=E501` on your PR (there are some operators with
>  missing whitespace), I'm not sure why flake8 doesn't catch that
>
>* no single line `ifs` please, e.g. if not "seen-days" in env: env["seen-days"] = set()

* these things fixed
>
>If you make this two PR (list and format) I believe we can merge format a lot
>earlier and can remove `event_description`.

* I would love to, I'm no sure I know how to (first because I'd have to
  look it up), but also because I'm not sure how to untangle them, list
  is heavily dependent on the event.format, and if I change things there
  then I'll have to pull them back and such. if we need to move on one
  at a time then I'll make an event.format, and after that is merged
  make a list request
>
>That's it for now...
>
>On Mon, May 16, 2016 at 07:47:59PM -0700, Youngmoney wrote:
>> I have done the refactoring I want to do, I looked into the calendar but the names I can get from "event" is the wrong name (not the display name), so I have no added the calendar name.
>>
>> ---
>> You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/pimutils/khal/pull/410#issuecomment-219606051
>
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/khal/pull/410#issuecomment-219668275

geier commented:
Thank you for fixing everything I brought up and the big improvement you made to khal!
